### PR TITLE
Improve reasoning about empty strings in rewriter

### PR DESCRIPTION
**Pull request by 4tXJ7f**
This commit adds several rewrites related to reasoning about empty
strings: it strengthens the rewrites for `(= "" _)` and it adds a
rewrite for `(str.replace "" y z)` that simplifies `y` if `(str.contains
"" y)` rewrites to a conjunction of empty-string equalities.

### CODE REVIEW

File: src/theory/strings/theory_strings_rewriter.cpp [hunk 1]
```diff
@@ -512,9 +519,17 @@ Node TheoryStringsRewriter::rewriteStrEqualityExt(Node node)
       {
         Node zero = nm->mkConst(Rational(0));
 
-        // (= "" (str.substr x n m)) ---> (<= (str.len x) n) if n >= 0 and m > 0
         if (checkEntailArith(ne[1], false) && checkEntailArith(ne[2], true))
         {
+          // (= "" (str.substr x 0 m)) ---> "" if m > 0
```
Comment (ajreynol) [inline src/theory/strings/theory_strings_rewriter.cpp]: I think you mean (= x "") ?

Comment (4tXJ7f) [reply src/theory/strings/theory_strings_rewriter.cpp]: Yup, fixed :)

File: src/theory/strings/theory_strings_rewriter.cpp [hunk 2]
```diff
@@ -2514,6 +2534,21 @@ Node TheoryStringsRewriter::rewriteReplace( Node node ) {
...
+      // be empty:
+      //
+      // (str.replace "" y z) ---> (str.replace "" (str.++ y1 ... yn)  z)
+      // if (str.contains "" y) ---> (and (= y1 "") ... (= yn ""))
+      if (node[0] == empty && allEqs)
+      {
+        std::vector<Node> emptyNodesList(emptyNodes.begin(), emptyNodes.end());
+        Node nn1 = mkConcat(STRING_CONCAT, emptyNodesList);
+        Node res = nm->mkNode(kind::STRING_STRREPL, node[0], nn1, nn2);
+        return returnRewrite(node, res, "rpl-emp-cnts-substs");
```
Comment (ajreynol) [inline src/theory/strings/theory_strings_rewriter.cpp]: Do we ensure that res != node here?  What if y is a CONCAT?

Comment (4tXJ7f) [reply src/theory/strings/theory_strings_rewriter.cpp]: Good point, done.

File: src/theory/strings/theory_strings_rewriter.cpp [hunk 3]
```diff
@@ -2504,6 +2520,10 @@ Node TheoryStringsRewriter::rewriteReplace( Node node ) {
             emptyNodes.insert(c[0]);
           }
         }
+        else
```
Comment (ajreynol) [inline src/theory/strings/theory_strings_rewriter.cpp]: Should we also set allEqs to false if we are EQUAL but neither side is empty?  Perhaps this should be allEmptyEqs?

Comment (4tXJ7f) [reply src/theory/strings/theory_strings_rewriter.cpp]: Yeah, good point. Done.

**Pull request merged**